stream/distributedlog/core/src/test/java/org/apache/distributedlog/TestTxnId.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.distributedlog;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.proto.BookieServer;
+import org.apache.bookkeeper.stats.NullStatsProvider;
+import org.apache.distributedlog.api.LogWriter;
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Test Cases for RollLogSegments.
+ */
+public class TestTxnId extends TestDistributedLogBase {
+    private static final Logger logger = LoggerFactory.getLogger(TestRollLogSegments.class);
+
+    @Test(timeout = 600000)
+    public void testRecoveryAfterBookieCrash() throws Exception {
+        String name = "txnid-after-crash";
+        DistributedLogConfiguration conf = new DistributedLogConfiguration()
+            .setEnsembleSize(5)
+            .setWriteQuorumSize(5)
+            .setAckQuorumSize(5);
+
+        long entryId = 0;
+        List<BookieServer> extraBookies = new ArrayList<>();
+        try {
+            extraBookies.add(startExtraBookie());
+            extraBookies.add(startExtraBookie());
+
+            try (BKDistributedLogManager dlm = (BKDistributedLogManager) createNewDLM(conf, name);
+                 BKAsyncLogWriter writer = dlm.startAsyncLogSegmentNonPartitioned()) {
+                writer.write(DLMTestUtil.getLogRecordInstance(1, 100000)).join();
+                writer.write(DLMTestUtil.getLogRecordInstance(2, 100000)).join();
+
+                extraBookies.forEach(b -> b.shutdown());
+
+                writer.write(DLMTestUtil.getLogRecordInstance(3, 100000));
+                Thread.sleep(1000);
```

Comment:
Is there any safer (not flaky) way of waiting?

Reply:
I can try. This sleep exists, to give time for for the write to error and the error handling to try to create a new ledger and then fail and mark the current log segment as null. Looks like it's exposed.